### PR TITLE
CreateHifu GraphQL Mutation

### LINKED ISSUE #21: CreateRoute Mutation
## As a frontend application, I will post a routes information to an mutation endpoint, so that I can save the user's route information.

The mutation will follow the following schema:

```json

mutation{
  createRoute( <submitted fields> ){
<return fields>
  }
}

```

### LINKED ISSUE #22: Create Types
### To make any endpoints we first need to create the following graphql types:
Required for user story: #21 
  - [ ] User Type
  - [ ] Route Type
  - [ ] Waypoint Type
  - [ ] Contact Type

### LINKED ISSUE #42: Create InputTypes
Our mutations require Input types for each of our tables

### PR DESCRIPTION
**Pull request by brian-greeson**
# Merge to Master PR

#### Closes: #21, 42

- [x] All test passing, Test percentage: 98%
- [x] Runs in local dev environment
- [x] PR Documented, Ready for Review

## Type of change made

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Test

### What's this PR do? (Detailed Description)

  Added a GraphQL endpoint `createHifu` which allows you to post an entire record {user,contact,route, and waypoints} in one post.  Exposes the created User back to the requester.

### Created (Classes, Methods, Migrations, etc.)

- Endpoints
   - `createHifu`: example
```graphql
 mutation{
        createHifu(
          user: {
          name:  String
          email: String <REQUIRED>
          phone: String
          address: String
          age: Int
          race: String
          gender: String
          satTrackerAddress: String
          bloodType: String
          allergies: String
          medicalConditions: String
          heightCM: Int
          weightKG: Int
          contact: {
            name: String
            email: String
            phone: String
          }
          route: {
            startTime: String
            endTime: String <REQUIRED>
            activity: String
            partySize: Int
            notes: String
            waypoints: [
              {
                latitude: Float <REQUIRED>
                longitude: Float <REQUIRED>
              }
            ]
          }
        }
        ){
          user{
            <All User Fields Exposed>
           }
           contact{
           <All Contact Field Exposed>
           }
           route{
           <All Route Fields Exposed>
           waypoints[
           {
           <All Waypoint Fields Exposed>
           }
          ]}
        }
      }
```
- Migrations
   - Removed start/end _time columns of type `string`
   - Added start/end_time columns of type `datetime`
- GraphQL Types
   - User, Contact, Route, and Waypoint Input types
   - User, Contact, Route, and Waypoint Query types

### Updated (Classes, Methods, Models, etc.)

N/A

### Technical Debt Added (What is not being tested? what does this PR break?)

- No sad path testing. 
- Returned Types not tested

### Notes

Data loss migrations, **Please reseed your local DB**

### CODE REVIEW

File: app/graphql/types/Input/route_input_type.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module Types
+  class Input::RouteInputType < Input::BaseInput
+    description "Attributes for creating or updating a route"
+    argument :userEmail, Integer, required: false
```
Comment (kristastadler) [inline app/graphql/types/Input/route_input_type.rb]: Should userEmail be an Integer? Elsewhere it's a string and that seems to make sense.

File: app/graphql/types/Input/route_input_type.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    argument :startTime, String, required: false
```
Comment (kristastadler) [inline app/graphql/types/Input/route_input_type.rb:5]: Didn't you change startTime and endTime to be DateTime objects?

File: app/graphql/types/route_type.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+module Types
+  class RouteType < Types::BaseObject
+    field :id, ID, null: false
+    field :user_id, Integer, null: false
+    field :start_time, String, null: true
```
Comment (kristastadler) [inline app/graphql/types/route_type.rb:5]: Same question -- should start_time and end_time be DateTime objects?

File: spec/graphql/types/mutation_type_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      expect(user.route.end_time).to eq(expected_user.route.end_time)
+      expect(user.route.activity).to eq(expected_user.route.activity)
+      expect(user.route.party_size).to eq(expected_user.route.party_size)
+      expect(user.route.notes).to eq(expected_user.route.notes)
+      expect(Waypoint.all.count).to eq(waypoints.count)
+      expect(user.route.waypoints.first.latitude).to eq(waypoints[0].latitude)
+      expect(user.route.waypoints.first.longitude).to eq(waypoints[0].longitude)
+      expect(user.route.waypoints.first.eta).to eq(waypoints[0].eta)
+      expect(user.route.waypoints.first.next.latitude).to eq(waypoints[1].latitude)
+      expect(user.route.waypoints.first.next.next.latitude).to eq(waypoints[2].latitude)
```
Comment (kristastadler) [inline spec/graphql/types/mutation_type_spec.rb:103]: Nice job with the testing on this!

**Comment by kristastadler** [comment]
This is a lot of functionality -- looks great!

File: app/graphql/types/Input/route_input_type.rb (hunk 2 above)
Comment (brian-greeson) [reply app/graphql/types/Input/route_input_type.rb:5]: They are datetime object in the database, but we want to allow the FE to post them as plain strings.  we can parse them into datetimes when we bring them in.  That will be part of sad path testing.

File: app/graphql/types/Input/route_input_type.rb (hunk 1 above)
Comment (brian-greeson) [reply app/graphql/types/Input/route_input_type.rb]: Excellent catch.  This was not being used yet (no authorization implemented), that why the test didn't catch it.

Comment (kristastadler) [reply app/graphql/types/Input/route_input_type.rb]: Do we need to fix it before merging this PR or do you want to add it to a bug list?

File: db/migrate/20200721231340_add_columnstarttime_to_routes.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class AddColumnstarttimeToRoutes < ActiveRecord::Migration[6.0]
```
Comment (margoflewelling) [inline db/migrate/20200721231340_add_columnstarttime_to_routes.rb:1]: I debated this when doing the original migration, string vs date time. I wasn't sure how the data from Megan would be formatted but I think date time is the right call

File: spec/graphql/types/mutation_type_spec.rb (hunk 4 above)
Comment (margoflewelling) [reply spec/graphql/types/mutation_type_spec.rb:103]: Second that ^. Very thorough.

File: db/migrate/20200721231340_add_columnstarttime_to_routes.rb (hunk 5 above)
Comment (brian-greeson) [reply db/migrate/20200721231340_add_columnstarttime_to_routes.rb:1]: You were definitely thinking along the right lines, but we can parse whatever we get from the FE into datetimes.  I think the datetimes will be much easier to work with in any business logic

**Comment by brian-greeson** [comment]
Github is not picking up the passing build from travis. Tried convert to draft and then back to normal PR, no luck.  Resubmitting PR seems like the only option

**Pull request merged**